Restructure (split gRPC and server code, better naming, Makefile)

sskaur opened the pull request:
These commits reorganize existing code but do not attempt to add any functionality.

@idryomov

sskaur committed:
file renaming

This commit reorganizes and renames files. It also adds
a Makefile to simplify protobuf compilation and use of
other basic commands like starting the server.

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

PepperJo commented on control/cli.py:
```diff
@@ -132,7 +132,7 @@ def connect(self, nvme_config):
         self._stub = pb2_grpc.NVMEGatewayStub(channel)
         # Set up logging
-        self._logger = nvme_config.logger
+        self._logger = settings.logger
```
It seems the logger does not have any interaction with any settings, so maybe move it out of the settings class.

PepperJo reviewed (changes requested):
Thanks Sandy for taking the time to split the PR.
I only have one minor item otherwise LGTM.

trociny commented on README.md:
```diff
@@ -50,7 +49,7 @@ This daemon runs as root. It provides the ability to export existing RBD images
 7. Start the gateway server daemon:
 		
-		$ python3 nvme_gw_server.py [-c config_filename]
+		$ python3 -m control [-s settings_file]
```
nit: Do we really need the server (and cli) as a module? Starting application like this looks a little weird to me. It is not that I care much but anyway...

PepperJo replied:
Yes, I guess the server and cli don't really need to be modules since they have no reuse anyway (i.e. will not be imported anywhere else). Or is there any other reason you made them modules @sskaur ? I mean we can still have them in their own directories without them being modules.

sskaur replied:
I had them as modules because I thought it might make it easier to import for finer grained unit tests in the future. There's also `make run` for starting the server, but I see your point with cli. I could pull out their respective argument parsing (in each __main__) and move them into scripts to run at the top level?

trociny replied:
Good point about unit tests.

`make run` is good for development but not for production.

Actually I am fine with any solution right now as it may be improved later. Eventually I suppose we will need some automatic "entry point" script creation like [1].

[1] https://setuptools.pypa.io/en/latest/userguide/quickstart.html#entry-points-and-automatic-script-creation

trociny commented on README.md:
```diff
@@ -69,14 +68,14 @@ Run the tool with the -h flag to see a list of available commands:
-			      			Path to config file
+    -h, --help            show this help message and exit
+    -s SETTINGS_FILE, --settings_file SETTINGS_FILE
+                          Path to settings file
```
`--settings_file` looks weird to me. I understand you renamed config to settings to avoid confusion with the persistent config, but I don't remember I have seen an application that has an option with such a name. Usually it is `--config`. If we really want to avoid confusion with the persistent config I would rather rename the persistent config to something else. Like `nvme.db`?

PepperJo replied:
@trociny I'm not sure I quite understand what you mean by `nvme.db` it's not necessarily about the file name it is about avoiding confusion in the code in respect to variable names and class names. Alternatively we can try to find another name for the persistent config.

trociny replied:
I am talking about another name for the persistent config. I.e. `DB` instead of `PersistenConfig`, and `nvme.db` object name instead of  `nvme.config`.

PepperJo replied:
Got it. `DB` is a bit too generic for my taste. Maybe something like `GWState` or `GroupState`? Any other ideas?

idryomov replied:
For the internal class names we can pick literally any name.  For the RADOS object name, in the iSCSI gateway it is actually configurable (in what is here referred to as a "settings file").

PepperJo replied:
The same here, the group file in the settings file (implicitly) defines the RADOS object name:
`self.omap_name = f"nvme.{gateway_group}.config"`

PepperJo replied:
The reason I initially asked to change from config to something else was that the code was confusing because there were two config types floating around and it was not always obvious which was which. So the idea was to give one a different name so we can clearly differentiate.

idryomov replied:
> The same here, the group file in the settings file (implicitly) defines the RADOS object name: `self.omap_name = f"nvme.{gateway_group}.config"`

Not quite, in the iSCSI gateway the entirety of the object name is up to the user -- they could make it "foobar".  But then, the iSCSI gateway doesn't have a concept of a gateway group so it doesn't need any sort of structure in the name.

BTW is the character set for the gateway group name constrained?  If not, it should be, to avoid weird object names if nothing else.

> The reason I initially asked to change from config to something else was that the code was confusing because there were two config types floating around and it was not always obvious which was which. So the idea was to give one a different name so we can clearly differentiate.

Makes sense, but I agree with Mykola.  The plaintext config file and related CLI options should continue to be termed config(uration), have `.conf` or `.config` extension, etc.  Let's rename the persistent "config" bits instead.

trociny replied:
> Got it. `DB` is a bit too generic for my taste. Maybe something like `GWState` or `GroupState`? Any other ideas?

`DB` was just an example. I am fine with any other name.

sskaur committed:
split grpc and server code

This commit splits gRPC API away from server code
that handles SPDK startup and monitoring.

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

sskaur committed:
move logging out of config

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

sskaur committed:
match RPC and cli commands

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

sskaur commented:
I changed `PersistentConfig` to `GatewayState` and updated other things (partially from discussion in #14). I believe it is ready for another look. @PepperJo @idryomov @trociny

trociny commented on README.md:
```diff
@@ -8,21 +8,20 @@ This daemon runs as root. It provides the ability to export existing RBD images
+		$ make setup
 		
-2. Modify the config file (default nvme_gw.config) to reflect the IP/ Port where the server can be reached:
+2. Modify the config file (default main.conf) to reflect the IP/ Port where the server can be reached:
```
I thought I had mentioned it in my previous review but it seems I forgot, sorry. Why "main.conf" name for the config file? Usually the config file name has the application name, which make it easier to associate with the application (when it is in /etc dir).

sskaur replied:
What is your opinion of "gateway.conf"? Do you think it should also be reflected in the module name?

trociny replied:
"gateway" is too generic. Why not "ceph-nvmeof.conf"? If we decided to use "ceph-nvmeof" for the project/repository name then it looks like a good name for the config.

As for the module name, although I don't think I am very happy with the current "control" name, but it is rather internal thing so I don't mind much. Eventually I believe users will not need to invoke the server as `python3 -m control`. There will be either a wrapper script or a container image with more meaningful name I suppose.

trociny commented on control/grpc.py:
```diff
@@ -209,10 +206,10 @@ def nvmf_subsystem_add_ns(self, request, context=None):
-    def nvmf_subsystem_remove_ns(self, request, context=None):
-        """Removes a given namespace from a given subsystem."""
+    def delete_namespace(self, request, context=None):
+        """Deletes a namespace from a subsystem."""
```
may be we should use "create-delete" and "add-remove" pairs?

PepperJo replied:
This would also align with the SPDK rpc API

PepperJo commented on control/cli.py:
```diff
@@ -31,12 +32,12 @@ class Parser:
             "-c",
-            "--config",
-            default="nvme_gw.config",
+            "--conffile",
```
I prefer `--config` but no stopper. Up to you.

PepperJo commented on control/cli.py:
```diff
@@ -282,18 +275,18 @@ def delete_host(self, args):
     ])
     def create_listener(self, args):
-        """Adds a listener at a particular TCP/IP address for a given subsystem."""
+        """Creates a listener for a subsystem at a given TCP/IP address."""
```
replace `TCP/IP address` with `IP/Port` -> e.g. RoCE/RDMA does not use TCP

PepperJo commented on control/cli.py:
```diff
@@ -307,29 +300,30 @@ def create_listener(self, args):
     ])
     def delete_listener(self, args):
-        """Deletes a listener at a particular TCP/IP address for a given subsystem."""
+        """Deletes a listener from a subsystem at a given TCP/IP address."""
```
see above

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+                json_req = json_format.MessageToJson(
+                    request, preserving_proto_field_name=True)
+                self.gateway_state.add_listener(request.nqn,
+                                                    request.gateway_name,
```
indent

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+                json_req = json_format.MessageToJson(
+                    request, preserving_proto_field_name=True)
+                self.gateway_state.add_host(request.subsystem_nqn,
+                                                request.host_nqn,
```
indent

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+        return pb2.req_status(status=ret)
+
+    def delete_listener(self, request, context=None):
+        """Deletes a listener from a subsystem at a given TCP/IP address."""
```
IP/Port

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+        return pb2.req_status(status=ret)
+
+    def create_listener(self, request, context=None):
+        """Creates a listener for a subsystem at a given TCP/IP address."""
```
`IP/Port`

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+            # Update gateway state
+            try:
+                self.gateway_state.delete_listener(request.nqn,
+                                                       request.gateway_name,
```
ident

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+                json_req = json_format.MessageToJson(
+                    request, preserving_proto_field_name=True)
+                self.gateway_state.add_subsystem(request.subsystem_nqn,
+                                                     json_req)
```
indent

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+                json_req = json_format.MessageToJson(
+                    request, preserving_proto_field_name=True)
+                self.gateway_state.add_namespace(request.subsystem_nqn,
+                                                     str(nsid), json_req)
```
indent

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+            # Update gateway state
+            try:
+                self.gateway_state.delete_namespace(request.subsystem_nqn,
+                                                        str(request.nsid))
```
indent

PepperJo commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+            # Update gateway state
+            try:
+                self.gateway_state.delete_host(request.subsystem_nqn,
+                                                   request.host_nqn)
```
indent

PepperJo reviewed (changes requested):
Thanks for the changes Sandy. Just some minor things otherwise LGTM.

sskaur committed:
variable renaming and comments

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

PepperJo reviewed (approved):
LGTM

trociny reviewed (approved):
LGTM

oritwas commented on COPYING:
```diff
@@ -4,7 +4,7 @@ Source: http://github.com/ceph/ceph-nvmeof
 License: LGPL-3.0-or-later (see LICENSE)
 
-Files: proto/nvme_gw.proto
+Files: control/proto/gateway.proto
```
I would keep the folder with protobuf definition at the highest level as it may be used to generate API externally.
For example, we will want to generate golang artifacts to use in Kubernetes for the CSI driver we will write

oritwas commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+	@python3 -m grpc_tools.protoc \
+			--proto_path=./$(MODULE)/proto \
+			--python_out=./$(MODULE)/proto \
+			--grpc_python_out=./$(MODULE)/proto \
```
The generated grpc file should be stored not with the model but in a separate folder, we will probably make it a package to be used in other projects/repos

sskaur replied:
For now I've placed the generated grpc files in the `generated` folder within `control` so as to not complicate relative imports, but this is flexible.

idryomov commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+			--python_out=./$(MODULE)/proto \
+			--grpc_python_out=./$(MODULE)/proto \
+			./$(MODULE)/proto/*.proto
+	@sed -E -i 's/^import.*_pb2/from . \0/' ./$(MODULE)/proto/*.py
```
Nit: it doesn't look like extended regexp syntax is actually used here, so `-E` is redundant.

idryomov commented on control/__main__.py:
```diff
@@ -0,0 +1,36 @@
+    logger = logging.getLogger(__name__)
+    if not os.path.isfile(args.config):
+        logger.error(f"Config file {args.config} not found.")
+        raise FileNotFoundError
```
This wouldn't catch things like `ceph-nvmeof.conf` file being there but inaccessible due to lacking permissions, etc.  I suspect the reason this check was added is that `configparser.read()` silently ignores files that it can't open, by design.  Since the gateway has a single config file location, a better fix would be to change `GatewayConfig` to use `configparser.read_file()` instead of `configparser.read()` and drop this check.

idryomov commented on control/cli.py:
```diff
@@ -178,56 +168,59 @@ def delete_bdev(self, args):
-            self.logger.info(f"Deleted namespace {delete_req.nsid}: {ret}")
+            ret = self.stub.remove_namespace(req)
+            self.logger.info(
+                f"Removed namespace {args.nsid} from {args.subnqn}:" +
```
Nit: implied line continuation in parentheses, etc is preferred in PEP-8.

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+        """Creates a bdev from an RBD image."""
+
+        self.logger.info(
+            f"Received request to create bdev {request.bdev_name} from" +
```
ditto

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+                if not request.traddr:
+                    traddr = self.config.get("gateway", "addr")
+                    if not traddr:
+                        raise Exception("config.gateway_addr option is not set")
```
This error string is out of date.

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+                if not request.traddr:
+                    traddr = self.config.get("gateway", "addr")
+                    if not traddr:
+                        raise Exception("config.gateway_addr option is not set")
```
ditto

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+                    f"Error persisting create_bdev {bdev_name}: {ex}")
+                raise
+
+        return pb2.req_status(status=True)
```
This used to construct `pb2.bdev_info`, was it replaced with status because the returned bdev name is guaranteed to be the same as requested?  Just want to make sure I understand the reason.

sskaur replied:
Yes. Is there some case where the returned bdev name would not match?

idryomov replied:
It feels like there shouldn't be but that was my question to you ;)

sskaur replied:
As far as I can tell, it will always match. Returning the bdev name was a holdover from an earlier commit where supplying the name in the cli command was optional and SPDK would generate the name. We switched to requiring it to make it easier for [restore](https://github.com/ceph/ceph-nvmeof/pull/9#discussion_r806796622), although now it seems like we could handle it the way we handle nsid and make it optional for the user again. This could be a separate small PR.

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+                                  f" {request.subsystem_nqn}: {ex}")
+                raise
+
+        return pb2.req_status(status=ret)
```
Same here, this used to construct `pb2.subsystem_info`?

sskaur replied:
`pb2.subsystem_info` would pass back the request parameter for subsystem nqn (as opposed to one returned from the SPDK RPC call), so I thought it was especially not needed.

idryomov replied:
Does `nvmf_create_subsystem` SPDK RPC return subsystem NQN?  https://spdk.io/doc/jsonrpc.html gives the following example response:
```
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": true
}
```

sskaur replied:
No, it returns status of the RPC call.

idryomov replied:
What did you mean by "as opposed to one returned from the SPDK RPC call" then?

sskaur replied:
I meant that if the SPDK RPC call returned a subsystem nqn, this would be more useful to return back from the gRPC than the request nqn. However, as with `rbd_bdev_create` and the returned bdev name, it would still not be needed. Therefore a case where the request nqn is returned is especially not needed. Sorry for being unclear earlier.

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,445 @@
+            nsid = self.spdk_rpc.nvmf.nvmf_subsystem_add_ns(
+                self.spdk_rpc_client,
+                nqn=request.subsystem_nqn,
+                bdev_name=request.bdev_name
```
`nsid` is no longer passed through to SPDK here, but it is still defined in the protocol as an optional request field.  Why is that?

sskaur replied:
It should still pass to SPDK at this point purposes of restore. Removing it was a mistake.

idryomov commented on control/state.py:
```diff
@@ -94,16 +95,16 @@ class OmapPersistentConfig(PersistentConfig):
 
-        gateway_group = self.nvme_config.get("config", "gateway_group")
+        gateway_group = self.config.get("gateway", "group")
         self.omap_name = f"nvme.{gateway_group}.config" if gateway_group else "nvme.config"
```
Should the RADOS object name be changed accordingly, e.g. to `nvmeof.{gateway_group}.state`/`nvmeof.state`?

idryomov commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+grpc:
+	@python3 -m grpc_tools.protoc \
+			--proto_path=./proto \
+			--python_out=./control/generated \
```
I don't think this would work without a pre-created directory (e.g. on a fresh checkout).

idryomov commented on control/grpc.py:
```diff
@@ -0,0 +1,446 @@
+    """
+
+    def __init__(self, config, gateway_state, spdk_rpc, spdk_rpc_client):
+
```
Is this missing `logging.basicConfig(level=logging.DEBUG)` like `server.py` used to?

sskaur replied:
`logging.basicConfig(level=logging.DEBUG)` configures the root logger that `server.py`, `state.py`, and `grpc.py`'s loggers inherit from so setting it once in `server.py` was sufficient for setting the log level. I changed this to set up the root logger in `main.py` to be more straightforward and because it makes more sense to set up more advanced logging features there.

idryomov commented on control/state.py:
```diff
@@ -94,16 +95,16 @@ class OmapPersistentConfig(PersistentConfig):
-        self.nvme_config = nvme_config
-        self.logger = nvme_config.logger
+        self.config = config
+        self.logger = logging.getLogger(__name__)
```
Same here, otherwise nothing below `WARNING` would appear in the log?

idryomov reviewed (commented):
Otherwise LGTM!
